// @turbo/core/lib/emitter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class EmitterBase {
    constructor() {
        this.map = {};
        this.mapOnce = {};
    }
    once(name, func) {
        if (this.mapOnce[name]) {
            this.mapOnce[name].add(func);
        }
        else {
            this.mapOnce[name] = new Set([func]);
        }
    }
    on(name, func) {
        if (this.map[name]) {
            this.map[name].add(func);
        }
        else {
            this.map[name] = new Set([func]);
        }
    }
    off(name, func) {
        if (this.map[name]) {
            this.map[name].delete(func);
            if (this.map[name].size === 0) {
                delete this.map[name];
            }
        }
        if (this.mapOnce[name]) {
            this.mapOnce[name].delete(func);
            if (this.mapOnce[name].size === 0) {
                delete this.mapOnce[name];
            }
        }
    }
    clear(name) {
        delete this.map[name];
    }
    fire(name, event) {
        if (this.map[name]) {
            for (const func of this.map[name]) {
                func(event);
            }
        }
        if (this.mapOnce[name]) {
            for (const func of this.mapOnce[name]) {
                func(event);
            }
            delete this.mapOnce[name];
        }
    }
}
exports.EmitterBase = EmitterBase;
