// turbo/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
process.on("uncaughtException", e => {
    console.error(e.message);
    console.error(e.stack);
    process.exit(1);
});
const yargs_1 = __importDefault(require("yargs"));
const start_1 = require("./commands/start");
const component_1 = require("./commands/component");
const eval_1 = require("./commands/eval");
const daemon_1 = require("./commands/daemon");
const ls_1 = require("./commands/ls");
const kill_1 = require("./commands/kill");
const core_1 = require("@turbo/core");
const turbo_1 = require("./turbo");
function getOptions(argv) {
    return {
        sessionId: argv.session || undefined,
    };
}
function makeTurbo(argv) {
    return turbo_1.getTurbo(getOptions(argv));
}
function run() {
    yargs_1.default
        .usage(core_1.LOGO)
        .help()
        .option("session", {
        describe: "the desired session id",
        alias: ["s"],
    })
        .command(["start", "$0"], "start a turbo session", {}, argv => {
        const turbo = makeTurbo(argv);
        start_1.start(turbo);
    })
        .command(["component <name>", "comp"], "start a turbo component", {}, argv => {
        const turbo = makeTurbo(argv);
        const name = argv.name;
        return component_1.component(turbo, name);
    })
        .command(["eval <expr>"], "evaluate an expression", {}, argv => {
        const turbo = makeTurbo(argv);
        const expr = argv.expr;
        return eval_1.evaluate(turbo, expr);
    })
        .command(["ls"], "list all running sessions", {}, argv => {
        const turbo = makeTurbo(argv);
        return ls_1.ls(turbo);
    })
        .command(["kill"], "kill a turbo session", {}, argv => {
        const turbo = makeTurbo(argv);
        return kill_1.kill(turbo);
    })
        .command("daemon", "start a turbo daemon", {}, argv => {
        const turbo = makeTurbo(argv);
        return daemon_1.daemon(turbo);
    }).argv;
}
exports.run = run;
