// turbo/lib/renderer/box.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
exports.Box = react_1.default.forwardRef(function InternalBox(props, ref) {
    let minHeight;
    let minWidth;
    const flexDirection = props.direction || "row";
    if (flexDirection === "row" || flexDirection === "row-reverse") {
        minHeight =
            typeof props.minHeight !== "undefined" ? props.minHeight : 1;
    }
    else {
        minWidth = typeof props.minWidth !== "undefined" ? props.minWidth : 1;
    }
    const style = {
        flexGrow: props.grow,
        flexShrink: props.shrink !== undefined ? props.shrink : 1,
        flexBasis: props.basis,
        flexDirection: flexDirection,
        alignItems: props.alignItems,
        justifyContent: props.justify,
        height: props.height,
        width: props.width,
        minHeight: minHeight,
        maxHeight: props.maxHeight,
        minWidth: minWidth,
        maxWidth: props.maxWidth,
        marginTop: props.marginTop,
        marginBottom: props.marginBottom,
        marginLeft: props.marginLeft,
        marginRight: props.marginRight,
        paddingTop: props.paddingTop,
        paddingBottom: props.paddingBottom,
        paddingLeft: props.paddingLeft,
        paddingRight: props.paddingRight,
    };
    const contentProps = {
        wrap: props.wrap,
        drawOffsetTop: props.drawOffsetTop,
        drawOffsetLeft: props.drawOffsetLeft,
        drawOverflow: props.drawOverflow,
        color: props.color,
        bg: props.bg,
        onClick: props.onClick,
        onMouse: props.onMouse,
        style,
        ref,
    };
    const element = react_1.default.createElement("div", contentProps, props.children);
    return element;
});
